sql/schema/v1-core.sql

```diff
@@ -620,77 +607,133 @@ REFERENCING NEW TABLE AS new_table
 FOR EACH STATEMENT
 EXECUTE FUNCTION after_v1_concurrency_slot_insert_function();
 
-CREATE OR REPLACE FUNCTION after_v1_concurrency_slot_delete_function()
-RETURNS trigger AS $$
+CREATE OR REPLACE FUNCTION cleanup_workflow_concurrency_slots(
+    p_strategy_id BIGINT,
+    p_workflow_version_id UUID,
+    p_workflow_run_id UUID
+) RETURNS VOID AS $$
+DECLARE
+    v_sort_id INTEGER;
 BEGIN
-    -- When v1_concurrency_slot is DELETED, we add it to the completed_child_strategy_ids on the parent, but only
-    -- when it's NOT a backoff retry. Backoff retries will continue to consume a workflow concurrency slot.
-    WITH parent_slot AS (
+    -- Get the sort_id for the specific workflow concurrency slot
+    SELECT sort_id INTO v_sort_id
+    FROM v1_workflow_concurrency_slot
+    WHERE strategy_id = p_strategy_id
+      AND workflow_version_id = p_workflow_version_id
+      AND workflow_run_id = p_workflow_run_id;
+
+    -- Acquire an advisory lock for the strategy ID
+    -- There is a small chance of collisions but it's extremely unlikely
+    PERFORM pg_advisory_xact_lock(1000000 * p_strategy_id + v_sort_id);
+
+    WITH final_concurrency_slots_for_dags AS (
+        -- If the workflow run id corresponds to a DAG, we get workflow concurrency slots
+        -- where NONE of the tasks in the associated DAG have v1_task_runtimes or v1_concurrency_slots
         SELECT
-            cs.workflow_id,
-            cs.workflow_version_id,
-            cs.workflow_run_id,
-            cs.strategy_id,
-            cs.parent_strategy_id
+            wcs.strategy_id,
+            wcs.workflow_version_id,
+            wcs.workflow_run_id
         FROM
-            deleted_rows cs
+            v1_workflow_concurrency_slot wcs
         JOIN
-            v1_task t ON t.id = cs.task_id AND t.inserted_at = cs.task_inserted_at
-        LEFT JOIN
-            v1_retry_queue_item rqi ON rqi.task_id = t.id AND rqi.task_inserted_at = t.inserted_at
+            v1_lookup_table lt ON wcs.workflow_run_id = lt.external_id AND lt.dag_id IS NOT NULL
         WHERE
-            cs.parent_strategy_id IS NOT NULL
-            AND rqi.task_id IS NULL
-    ), parent_slots_grouped AS (
+            wcs.strategy_id = p_strategy_id
+            AND wcs.workflow_version_id = p_workflow_version_id
+            AND wcs.workflow_run_id = p_workflow_run_id
+            AND NOT EXISTS (
```

Comment:
Given that we're processing singletons, I'm not too concerned about `NOT EXISTS` but will need to test